sklearn/_loss/_loss.pyx.tp

```diff
@@ -0,0 +1,1213 @@
+        """
+        pass
+
+    def _loss_gradient(
+        self,
+        Y_DTYPE_C[::1] y_true,          # IN
+        Y_DTYPE_C[::1] raw_prediction,  # IN
+        Y_DTYPE_C[::1] sample_weight,   # IN
+        G_DTYPE_C[::1] loss,            # OUT
+        G_DTYPE_C[::1] gradient,        # OUT
+        int n_threads=1
+    ):
+        """Compute loss and gradient of loss w.r.t raw_prediction.
+
+        Parameters
+        ----------
+        y_true : array of shape (n_samples,)
+            Observed, true target values.
+        raw_prediction : array of shape (n_samples,)
+            Raw prediction values (in link space).
+        sample_weight : array of shape (n_samples,) or None
+            Sample weights.
+        loss : array of shape (n_samples,) or None
+            A location into which the element-wise loss is stored.
+        gradient : array of shape (n_samples,)
+            A location into which the gradient is stored.
+        n_threads : int
+            Number of threads used by OpenMP (if any).
+
+        Returns
+        -------
+        loss : array of shape (n_samples,)
+            Element-wise loss function.
+
+        gradient : array of shape (n_samples,)
+            Element-wise gradients.
+        """
+        self._loss(y_true, raw_prediction, sample_weight, loss,
+                            n_threads)
+        self._gradient(y_true, raw_prediction, sample_weight, gradient,
+                      n_threads)
+        return np.asarray(loss), np.asarray(gradient)
+
+    def _gradient_hessian(
+        self,
+        Y_DTYPE_C[::1] y_true,          # IN
+        Y_DTYPE_C[::1] raw_prediction,  # IN
+        Y_DTYPE_C[::1] sample_weight,   # IN
+        G_DTYPE_C[::1] gradient,        # OUT
+        G_DTYPE_C[::1] hessian,         # OUT
+        int n_threads=1
+    ):
+        """Compute gradient and hessian of loss w.r.t raw_prediction.
+
+        Parameters
+        ----------
+        y_true : array of shape (n_samples,)
+            Observed, true target values.
+        raw_prediction : array of shape (n_samples,)
+            Raw prediction values (in link space).
+        sample_weight : array of shape (n_samples,) or None
+            Sample weights.
+        gradient : array of shape (n_samples,)
+            A location into which the gradient is stored.
+        hessian : array of shape (n_samples,)
+            A location into which the hessian is stored.
+        n_threads : int
+            Number of threads used by OpenMP (if any).
+
+        Returns
+        -------
+        gradient : array of shape (n_samples,)
+            Element-wise gradients.
+
+        hessian : array of shape (n_samples,)
+            Element-wise hessians.
+        """
+        pass
+
+
+{{for name, docstring, param, closs, closs_grad, cgrad, cgrad_hess, in class_list}}
+{{py:
+if param is None:
+    with_param = ""
+else:
+    with_param = ", self." + param
+}}
+
+cdef class {{name}}(cLossFunction):
```

Comment:
I guess `cython.final` be useful here? From [the docs](https://cython.readthedocs.io/en/latest/src/tutorial/pure.html):

> `@cython.final` terminates the inheritance chain by preventing a type from being used as a base class, or a method from being overridden in subtypes. This enables certain optimisations such as inlined method calls.

From the code, it removes some boilerplate [for attribute lookups](https://github.com/cython/cython/blob/f6c850d11bc57a432519afb7369bf0d08008be25/Cython/Compiler/ModuleNode.py#L2315-L2319) and [for type safety checks](https://github.com/cython/cython/blob/f6c850d11bc57a432519afb7369bf0d08008be25/Cython/Compiler/ModuleNode.py#L1540-L1561). (Searching for [`is_final_type`](https://github.com/cython/cython/search?q=is_final_type) and [`is_final_cmethod`](https://github.com/cython/cython/search?q=is_final_cmethod) in Cython's code-base are good entry-points to see how this gets made).
```suggestion
@cython.final
cdef class {{name}}(cLossFunction):
```

Reply:
For instance, this is the diff in the generated C code when marking the following ext type as `final`:

```python
from cython cimport final
@final # change this
cdef class MyExtType:
    cdef int an_attribute
    @final # this is not necessary
    cdef int a_method(self, int an_argument):
        return 1

```
```patch
diff --git a/a.old.c b/a.c
index f32344844..f093c1be0 100644
--- a/a.old.c
+++ b/a.c
@@ -1040,10 +1040,10 @@ struct __pyx_obj_1a_MyExtType;
 
 /* "a.pyx":3
  * from cython cimport final
- * 
+ * @final
  * cdef class MyExtType:             # <<<<<<<<<<<<<<
  *     cdef int an_attribute
- * 
+ *     @final
  */
 struct __pyx_obj_1a_MyExtType {
   PyObject_HEAD
@@ -1057,6 +1057,7 @@ struct __pyx_vtabstruct_1a_MyExtType {
   int (*a_method)(struct __pyx_obj_1a_MyExtType *, int);
 };
 static struct __pyx_vtabstruct_1a_MyExtType *__pyx_vtabptr_1a_MyExtType;
+static int __pyx_f_1a_9MyExtType_a_method(struct __pyx_obj_1a_MyExtType *, int);
 /* #### Code section: utility_code_proto ### */
 
 /* --- Runtime support code (head) --- */
@@ -1411,13 +1412,6 @@ static CYTHON_INLINE PyObject* __Pyx_PyObject_GenericGetAttrNoDict(PyObject* obj
 #define __Pyx_PyObject_GenericGetAttrNoDict PyObject_GenericGetAttr
 #endif
 
-/* PyObject_GenericGetAttr.proto */
-#if CYTHON_USE_TYPE_SLOTS && CYTHON_USE_PYTYPE_LOOKUP && PY_VERSION_HEX < 0x03070000
-static PyObject* __Pyx_PyObject_GenericGetAttr(PyObject* obj, PyObject* attr_name);
-#else
-#define __Pyx_PyObject_GenericGetAttr PyObject_GenericGetAttr
-#endif
-
 /* SetVTable.proto */
 static int __Pyx_SetVtable(PyTypeObject* typeptr , void* vtable);
 
@@ -2039,7 +2033,7 @@ static int __pyx_m_traverse(PyObject *m, visitproc visit, void *arg) {
 
 /* "a.pyx":6
  *     cdef int an_attribute
- * 
+ *     @final
  *     cdef int a_method(self, int an_argument):             # <<<<<<<<<<<<<<
  *         return 1
  */
@@ -2050,7 +2044,7 @@ static int __pyx_f_1a_9MyExtType_a_method(CYTHON_UNUSED struct __pyx_obj_1a_MyEx
   __Pyx_RefNannySetupContext("a_method", 0);
 
   /* "a.pyx":7
- * 
+ *     @final
  *     cdef int a_method(self, int an_argument):
  *         return 1             # <<<<<<<<<<<<<<
  */
@@ -2059,7 +2053,7 @@ static int __pyx_f_1a_9MyExtType_a_method(CYTHON_UNUSED struct __pyx_obj_1a_MyEx
 
   /* "a.pyx":6
  *     cdef int an_attribute
- * 
+ *     @final
  *     cdef int a_method(self, int an_argument):             # <<<<<<<<<<<<<<
  *         return 1
  */
@@ -2865,11 +2859,7 @@ static PyObject *__pyx_tp_new_1a_MyExtType(PyTypeObject *t, CYTHON_UNUSED PyObje
   allocfunc alloc_func = (allocfunc)PyType_GetSlot(t, Py_tp_alloc);
   o = alloc_func(t, 0);
   #else
-  if (likely(!__Pyx_PyType_HasFeature(t, Py_TPFLAGS_IS_ABSTRACT))) {
-    o = (*t->tp_alloc)(t, 0);
-  } else {
-    o = (PyObject *) PyBaseObject_Type.tp_new(t, __pyx_empty_tuple, 0);
-  }
+  o = (*t->tp_alloc)(t, 0);
   if (unlikely(!o)) return 0;
   #endif
   p = ((struct __pyx_obj_1a_MyExtType *)o);
@@ -2878,11 +2868,6 @@ static PyObject *__pyx_tp_new_1a_MyExtType(PyTypeObject *t, CYTHON_UNUSED PyObje
 }
 
 static void __pyx_tp_dealloc_1a_MyExtType(PyObject *o) {
-  #if CYTHON_USE_TP_FINALIZE
-  if (unlikely((PY_VERSION_HEX >= 0x03080000 || __Pyx_PyType_HasFeature(Py_TYPE(o), Py_TPFLAGS_HAVE_FINALIZE)) && Py_TYPE(o)->tp_finalize) && (!PyType_IS_GC(Py_TYPE(o)) || !_PyGC_FINALIZED(o))) {
-    if (PyObject_CallFinalizerFromDealloc(o)) return;
-  }
-  #endif
   (*Py_TYPE(o)->tp_free)(o);
 }
 
@@ -2902,7 +2887,7 @@ static PyType_Spec __pyx_type_1a_MyExtType_spec = {
   "a.MyExtType",
   sizeof(struct __pyx_obj_1a_MyExtType),
   0,
-  Py_TPFLAGS_DEFAULT|Py_TPFLAGS_HAVE_VERSION_TAG|Py_TPFLAGS_CHECKTYPES|Py_TPFLAGS_HAVE_NEWBUFFER|Py_TPFLAGS_BASETYPE,
+  Py_TPFLAGS_DEFAULT|Py_TPFLAGS_HAVE_VERSION_TAG|Py_TPFLAGS_CHECKTYPES|Py_TPFLAGS_HAVE_NEWBUFFER,
   __pyx_type_1a_MyExtType_slots,
 };
 #else
@@ -2937,7 +2922,7 @@ static PyTypeObject __pyx_type_1a_MyExtType = {
   0, /*tp_getattro*/
   0, /*tp_setattro*/
   0, /*tp_as_buffer*/
-  Py_TPFLAGS_DEFAULT|Py_TPFLAGS_HAVE_VERSION_TAG|Py_TPFLAGS_CHECKTYPES|Py_TPFLAGS_HAVE_NEWBUFFER|Py_TPFLAGS_BASETYPE, /*tp_flags*/
+  Py_TPFLAGS_DEFAULT|Py_TPFLAGS_HAVE_VERSION_TAG|Py_TPFLAGS_CHECKTYPES|Py_TPFLAGS_HAVE_NEWBUFFER, /*tp_flags*/
   0, /*tp_doc*/
   0, /*tp_traverse*/
   0, /*tp_clear*/
@@ -3240,7 +3225,7 @@ static int __Pyx_modinit_type_init_code(void) {
   #endif
   #if !CYTHON_COMPILING_IN_LIMITED_API
   if ((CYTHON_USE_TYPE_SLOTS && CYTHON_USE_PYTYPE_LOOKUP) && likely(!__pyx_ptype_1a_MyExtType->tp_dictoffset && __pyx_ptype_1a_MyExtType->tp_getattro == PyObject_GenericGetAttr)) {
-    __pyx_ptype_1a_MyExtType->tp_getattro = __Pyx_PyObject_GenericGetAttr;
+    __pyx_ptype_1a_MyExtType->tp_getattro = __Pyx_PyObject_GenericGetAttrNoDict;
   }
   #endif
   if (__Pyx_SetVtable(__pyx_ptype_1a_MyExtType, __pyx_vtabptr_1a_MyExtType) < 0) __PYX_ERR(1, 3, __pyx_L1_error)
@@ -3586,7 +3571,7 @@ if (!__Pyx_RefNanny) {
 
   /* "a.pyx":1
  * from cython cimport final             # <<<<<<<<<<<<<<
- * 
+ * @final
  * cdef class MyExtType:
  */
   __pyx_t_1 = __Pyx_PyDict_NewPresized(0); if (unlikely(!__pyx_t_1)) __PYX_ERR(1, 1, __pyx_L1_error)
@@ -5231,16 +5216,6 @@ static CYTHON_INLINE PyObject* __Pyx_PyObject_GenericGetAttrNoDict(PyObject* obj
 }
 #endif
 
-/* PyObject_GenericGetAttr */
-#if CYTHON_USE_TYPE_SLOTS && CYTHON_USE_PYTYPE_LOOKUP && PY_VERSION_HEX < 0x03070000
-static PyObject* __Pyx_PyObject_GenericGetAttr(PyObject* obj, PyObject* attr_name) {
-    if (unlikely(Py_TYPE(obj)->tp_dictoffset)) {
-        return PyObject_GenericGetAttr(obj, attr_name);
-    }
-    return __Pyx_PyObject_GenericGetAttrNoDict(obj, attr_name);
-}
-#endif
-
 /* SetVTable */
 static int __Pyx_SetVtable(PyTypeObject *type, void *vtable) {
     PyObject *ob = PyCapsule_New(vtable, 0, 0);

```

Reply:
You could report those findings back to cython itself.